### PR TITLE
Freesurfer ROI Inputs

### PR DESCRIPTION
**Pull request by swashcap**
Asana task: https://app.asana.com/0/37131652076106/167594259597250

Related:
- MRN-Code/decentralized-laplacian-ridge-regression#6
- MRN-Code/freesurfer-parser#1

Todo (once reviewed):
- [x] Bump _freesurfer-parser_’s version and publish
- [x] Update _freesurfer-parser_’s version in _decentralized-laplacian-ridge-regression_’s _package.json_
- [x] Update  _decentralized-laplacian-ridge-regression_’s version and publish
- [x] Bump `laplacian-noise-ridge-regression`’s version in _packages/coinstac-common/src/decentralized-computations.json_

### CODE REVIEW

File: packages/coinstac-client-core/bin/coinstac-client-core [hunk 1]
```diff
@@ -3,9 +3,10 @@
```
Comment (swashcap) [inline packages/coinstac-client-core/bin/coinstac-client-core:3]: Just a couple changes on the client CLI to make it work. Not super important.

File: packages/coinstac-common/src/models/consortium.js [hunk 2]
```diff
@@ -53,6 +53,17 @@ class Consortium extends PouchDocument {
 Consortium.schema = Object.assign({}, PouchDocument.schema, {
   _id: joi.string().regex(/^[a-zA-Z0-9]+$/),
   activeComputationId: joi.string().min(1),
+
+  /**
+   * @todo A computation's inputs shouldn't be set on the consortium because
+   * this limits runs to one-at-a-time. Consider moving elsewhere, perhaps some
+   * sort of `ComputationRun` model.
+   */
+  activeComputationInputs: joi.array().items(
```
Comment (swashcap) [inline packages/coinstac-common/src/models/consortium.js:62]: Owners select inputs in the UI, then save them here. These are serialized into the local and remote doc for each run, so `activeComputationInputs` can be changed for the next run.

File: packages/coinstac-common/src/models/decentralized-computation.js [hunk 3]
```diff
@@ -25,14 +25,59 @@ class DecentralizedComputation extends Base {
...
    * @property {string} name
    * @property {string} url
    * @property {string} version
    */
   getComputationDocument() {
     const { name, repository: { url }, version } = this.serialize();
+    const doc = { name, url, version };
 
-    return { name, url, version };
+    /**
```
Comment (swashcap) [inline packages/coinstac-common/src/models/decentralized-computation.js:37]: These changes ensure a computation definition’s `inputs` are saved to the `DecentralizedComputation`’s document in the _computations_ database. It’s pretty rough: we should offer some sort of plugin for computations to save relevant info to their docs.

File: packages/coinstac-common/test/models/pipeline/plugins/inputs.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
+'use strict';
```
Comment (swashcap) [inline packages/coinstac-common/test/models/pipeline/plugins/inputs.js:1]: Test 👏 the 👏 things 👏

File: packages/coinstac-ui/app/render/state/ducks/consortia.js [hunk 5]
```diff
@@ -104,6 +105,134 @@ export const setActiveComputation = applyAsyncLoading(
 );
 
 /**
+ * Set inputs on a consortium's selected computation.
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/state/ducks/consortia.js:109]: :boom: Document design discussion here. We should use these ideas when expanding the idea of “inputs” in COINSTAC. (See #12)

File: packages/coinstac-common/src/models/consortium.js (hunk 2 above)
Comment (rssk) [reply packages/coinstac-common/src/models/consortium.js:62]: Cool 👍

File: packages/coinstac-common/src/models/pipeline/plugins/inputs.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+     * @param {LocalComputationResult} compResult
+     * @param {Object} hooks
+     * @param {Function} hooks.cancelRun
+     * @param {Function} hooks.forceSave
+     * @param {Function} hooks.isRunCancelled
+     * @returns {(LocalComputationResult|undefined)}
+     */
+    local(runInput, compResult, hooks) {
+      /**
+       * @todo: Determine why are a client's local info is stored on a
```
Comment (rssk) [inline packages/coinstac-common/src/models/pipeline/plugins/inputs.js:38]: hah, yeah :-/

File: packages/coinstac-common/test/models/pipeline/plugins/inputs.js (hunk 4 above)
Comment (rssk) [reply packages/coinstac-common/test/models/pipeline/plugins/inputs.js:1]: test all the things

**Comment by rssk** [comment]
Looks good! You have my blessing to merge when you get it cleaned up ??

[--- Commented from Asana.com
#commenter coins
---[aa]

**Comment by swashcap** [comment]
Noticing that the remote result document’s `data.iteration` is `100`, but the doc’s `history` array has a length of `199`. Very…odd.

This occurs when the computation is configured via a consortium input to max out at 100 runs. I wonder if the inputs plugin is causing unnecessary saves?

**Comment by swashcap** [comment]
Testing this feature is unfortunately rather involved as it requires coordinating with [MRN-Code/decentralized-laplacian-ridge-regression](https://github.com/MRN-Code/decentralized-laplacian-ridge-regression)’s feature branch. Here’s how to test:
1. `git clone git@github.com:MRN-Code/decentralized-laplacian-ridge-regression.git` somewhere _special_.
2. Check out the _feature/freesurfer-rio-inputs_ branch in both _coinstac_ and _decentralized-laplacian-ridge-regression_.
   - Run `npm i` in _decentralized-laplacian-ridge-regression_
   - Run `npm run bootstrap` in _coinstac_
3. Clean state:
   1. Start couchdb
   2. Run `npm run clean:db` in _coinstac/packages/coinstac-server-core_
   3. Run `bin/coinstac-server-core --nohang` in the same directory to set up the _consortia_ and _computations_ databases
   4. Create a sample consortium (you can use [coinstac-setup](https://github.com/swashcap/coinstac-setup) to help)
   5. Run `npm run clean-db` in _coinstac/packages/coinstac-ui_
4. Here’s a nice bash script:
   
   ``` bash
   #!/bin/bash
   
   COMP_DIR="$(node -e "console.log(require('os').tmpDir() + '/coinstac-server-core/computations')" | tr -d '\r')"
   ALGO="laplacian-noise-ridge-regression@2.0.3"
   
   mkdir -p ~/.coinstac/computations
   cd ~/.coinstac/computations
   rm -rf $ALGO
   mkdir $ALGO
   echo "module.exports = require('$HOME/Sites/decentralized-laplacian-ridge-regression')" > $ALGO/index.js
   
   mkdir -p $COMP_DIR
   cd $COMP_DIR
   rm -rf $ALGO
   mkdir $ALGO
   echo "module.exports = require('$HOME/Sites/decentralized-laplacian-ridge-regression')" > $ALGO/index.js
   ```
   
   **Notice the `$HOME/Sites` prefix: change this to match the location of your copy.**
   
   Run this to stub in the feature branch version of _decentralized-laplacian-ridge-regression_. (Any idea how to refactor that into bash functions?)
5. Boot up two clients (`node_modules/.bin/electron .` in _coinstac/packages/coinstac-ui_) and attempt to run!

**Comment by swashcap** [comment]
The ridge-regression’s iteration count can’t be changed in the UI. ??

[--- Commented from Asana.com
#commenter coins
---[aa]

**Comment by swashcap** [comment]
> The ridge-regression’s iteration count can’t be changed in the UI

This appears to be due to RPC-ing objects in Electron. The `Consortium` doesn’t have a default `activeComputationInputs` property. Adding this property in the render process doesn’t add it in the remote process, and it isn’t actually being saved to the database.

File: packages/coinstac-ui/app/render/components/consortium-single.js [hunk 7]
```diff
@@ -140,18 +140,10 @@ class ConsortiumSingle extends Component {
 
             if (type === 'number') {
               fieldProps.onChange = event => {
-                if (typeof event === 'number') {
```
Comment (swashcap) [inline packages/coinstac-ui/app/render/components/consortium-single.js]: `event` is no longer a number. It's always shaped like a DOM event, regardless of whether it is an actual event or an increment/decrement click.

File: packages/coinstac-common/src/models/consortium.js [hunk 8]
```diff
@@ -65,7 +65,7 @@ Consortium.schema = Object.assign({}, PouchDocument.schema, {
       joi.number(),
       joi.string()
     )
-  ),
+  ).default([]),
```
Comment (swashcap) [inline packages/coinstac-common/src/models/consortium.js:68]: Fixes the `activeComputationInput` RPC bug.

**Pull request merged**